packages/event-listener/src/index.ts

```diff
@@ -1,128 +1,135 @@
-  const targets: EventTarget[] = Array.isArray(target) ? target : [target];
-  type EventProps = [
-    name: string,
-    handler: EventListenerOrEventListenerObject | null,
-    options?: EventListenerOptions
-  ];
-  const props: Accessor<EventProps> =
-    typeof propsArray[0] === "function" ? propsArray[0] : () => propsArray as EventProps;
-  const add = (target: EventTarget) => {
-    targets.includes(target) || targets.push(target);
-    const [name, handler, options] = props();
-    if (name && handler !== undefined) {
-      target.addEventListener.apply(target, props());
-    }
-  };
-  const remove = (target: EventTarget) => {
-    targets.forEach((t, index) => t === target && targets.splice(index, 1));
-    target.removeEventListener.apply(target, props());
+
+// DOM Events
+export function createEventListener<
+  Target extends Window | Document | HTMLElement | MediaQueryList,
+  EventMap extends EventMapOf<Target>,
+  EventName extends keyof EventMap
+>(
+  target: MaybeAccessor<Many<Target>>,
+  eventName: EventName,
+  handler: EventHandler<EventMap, EventName>,
+  options?: boolean | AddEventListenerOptions
+): EventListenerReturn;
+
+// Custom Events
+export function createEventListener<
+  EventMap extends Record<string, Event>,
+  EventName extends keyof EventMap = keyof EventMap
+>(
+  target: MaybeAccessor<Many<EventTarget>>,
+  eventName: EventName,
+  handler: EventHandler<EventMap, EventName>,
+  options?: boolean | AddEventListenerOptions
+): EventListenerReturn;
+
+// Directive usage
+export function createEventListener(
+  target: MaybeAccessor<Many<EventTarget>>,
+  props: Accessor<EventListenerDirectiveProps>
+): EventListenerReturn;
+
+export function createEventListener(
+  target: MaybeAccessor<Many<EventTarget>>,
+  ...rest: any
+): EventListenerReturn {
+  const props: Accessor<EventListenerDirectiveProps> = rest.length > 1 ? () => rest : rest[0];
+
+  const toCleanup = createCallbackStack();
+  const start = () => {
+    if (!isClient) return;
+    toCleanup.execute();
+    const [eventName, handler, options] = props();
+    accessAsArray(target).forEach(target => {
+      if (!target) return;
+      target.addEventListener(eventName, handler, options);
+      toCleanup.push(() => target.removeEventListener(eventName, handler, options));
+    });
   };
-  // we need to directly add the event, otherwise we cannot dispatch it before the next effect runs
-  targets.forEach(add);
-  createEffect(previousProps => {
-    const currentProps = props();
-    if (previousProps !== currentProps) {
-      previousProps &&
-        targets.forEach(target => target.removeEventListener.apply(target, previousProps));
-      targets.forEach(add);
-    }
-    return currentProps;
-  }, props());
-  onCleanup(() => {
-    targets.forEach(remove);
-  });
-  return [add, remove];
-};
 
-export default createEventListener;
+  createEffect(start);
+  onCleanup(toCleanup.execute);
+
+  return [toCleanup.execute, start];
```

Comment:
This is not the same functionality, it contains [stop, start] (why not the other way around?) instead of [add, remove]. We should at least discuss this. Since we allow rebinding, using `null` as a handler already allows stopping and you can add a handler back later. On the other hand, the target can be an accessor as well and thus, add and remove are superfluous as well.

So I would probably vote for another option: to make the last event available as reactive signal, thus basically making this doing the work of rxjs' `fromEvent`.

Reply:
I've put [stop, start] in this order, because 1. stop will be more often used than start, and 2. you first need to first stop the events to be able to start them since they start automatically.
I switched add and remove because I needed an easy way to remove the event listeners, not by just passing a nop as a handler, but removing event listeners from elements.
I said that it has the same functionality because modifying the targets is still possible since the target arg is now reactive:
```ts
  const [targets, setTargets] = createSignal([
    document.getElementById("123"),
    document.getElementById("abc")
  ]);

  createEventListener(targets, "mouseenter", () => {});

  const add = (el: HTMLElement) => setTargets(p => [...p, el]);
  const remove = (el: HTMLElement) =>
    setTargets(p => {
      const copy = p.slice();
      copy.splice(p.indexOf(el), 1);
      return copy;
    });
```
This may be a bit lengthy, but it's a common pattern to use in Solid. Since signals are immutable, this is just how you handle arrays.
I've chosen this approach because it seemed more aligned with reactivity and solid patterns.
But I guess the same could be said about stop and start. Although this is a little bit different because these are supposed to be permanent actions, If you want to disable handling events temporarily, then putting that logic into a handler is fine.
So maybe we should move away from this too, and instead, rely on Solid's patterns again to provide solutions for users:
```ts
  const stop = createRoot(stop => {
    createEventListener(targets, "mouseenter", () => {});
    return stop;
  });
```
This way the space would be left for the `lastEvent` as a returned value:
```ts
const lastEvent = createEventListener(targets, "mouseenter", () => {});
```

Reply:
I would implement `remove` as

```ts
const remove = (el: HTMLElement) => setTargets(t => t.filter(target => target !== el));
// and this allows us also to remove/stop the events easily:
const clear = () => setTargets([]);
```
since we need a new reference for the array anyways to propagate the change.

@davedbase what's your opinion? I copied the "add" and "remove" methods from your initial implementation. I think we should rather make sure that w handle the target array reactively (do not needlessly remove and add the same listener to the same target).

If we want to add more methods, we could still use a `[lastEvent, methods]` pattern, but I think that's overengineering.